pkg/migration/types.go

```diff
@@ -0,0 +1,135 @@
+	// to complete a migration when executed in order.
+	Steps []Step `json:"steps,omitempty"`
+}
+
+// StepType is the type used to name a migration step
+type StepType string
+
+const (
+	// StepTypeApply denotes an apply step
+	StepTypeApply StepType = "Apply"
+	// StepTypeDelete denotes a delete step
+	StepTypeDelete StepType = "Delete"
+)
+
+// Step represents a step in the generated migration plan
+type Step struct {
+	// Name is the name of this Step
+	Name string `json:"name"`
+	// Type is the type of this Step.
+	// Can be one of Apply, Delete, etc.
+	Type StepType `json:"type"`
+	// Apply contains the information needed to run an StepTypeApply step.
+	// Must be set when the Step.Type is StepTypeApply.
+	Apply *ApplyStep `json:"apply,omitempty"`
+	// Delete contains the information needed to run an StepTypeDelete step.
+	// Must be set when the Step.Type is StepTypeDelete.
+	Delete *DeleteStep `json:"delete,omitempty"`
+}
+
+// ApplyStep represents an apply step in which an array of manifests
+// is applied from the filesystem.
+type ApplyStep struct {
+	// Files denotes the paths of the manifest files to be applied.
+	// The paths can either be relative or absolute.
+	Files []string `json:"files,omitempty"`
+}
+
+// DeleteStep represents a deletion step with options
+type DeleteStep struct {
+	// Options represents the options to be used while deleting the resources
+	// specified in Resources.
+	Options *DeleteOptions `json:"options,omitempty"`
+	// Resources is the array of resources to be deleted in this step
+	Resources []Resource `json:"resources,omitempty"`
+}
+
+// DeleteOptions represent options to be used during deletion of
+// a managed resource.
+type DeleteOptions struct {
+	// FinalizerPolicy denotes the policy to be used regarding
+	// the managed reconciler's finalizer
+	FinalizerPolicy *FinalizerPolicy `json:"finalizerPolicy,omitempty"`
+}
+
+// GroupVersionKind represents the GVK for an object's kind.
+// schema.GroupVersionKind does not contain json the serialization tags
+// for its fields, but we would like to serialize these as part of the
+// migration plan.
+type GroupVersionKind struct {
+	// Group is the API group for the resource
+	Group string `json:"group"`
+	// Version is the API version for the resource
+	Version string `json:"version"`
+	// Kind is the kind name for the resource
+	Kind string `json:"kind"`
+}
+
+type Resource struct {
+	// GroupVersionKind holds the GVK for the resource's type
+	// schema.GroupVersionKind is not embedded for consistent serialized names
+	GroupVersionKind `json:",inline"`
+	// Name is the name of the resource
+	Name string `json:"name"`
+}
+
+// Metadata holds metadata for an object read from a Source
+type Metadata struct {
+	// Path uniquely identifies the path for this object on its Source
+	Path string
+	// colon separated list of parent `Path`s for fan-ins and fan-outs
+	// Example: resources/a.yaml:resources/b.yaml
+	Parents string
```

Comment:
I see this field isn't used anywhere. Is it a leftover?

Reply:
This field is to be employed by the source and target implementations.